### PR TITLE
Add support for specifying private fields through a strawberry.type variable

### PR DESCRIPTION
**Pull request by lijok**
## Description
`strawberry.Private` is great for specifying private fields, however, does not play well with IDE's: https://github.com/strawberry-graphql/strawberry/issues/702

This PR adds supports for specifying which fields on a type should be made private, through a `private_fields` variable on `strawberry.type` decorator

Example:
```py
import strawberry

@strawberry.type(private_fields=["uid"])
class Post:
    uid: str
    title: str
    description: str
```

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/strawberry-graphql/strawberry/issues/702

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

File: strawberry/type.py [hunk 1]
```diff
@@ -97,13 +97,17 @@ def _process_type(
...
     name = name or to_camel_case(cls.__name__)
 
     wrapped = _wrap_dataclass(cls)
 
     interfaces = _get_interfaces(wrapped)
-    fields = _get_fields(cls)
+    if private_fields:
+        fields = [i for i in _get_fields(cls) if i.origin_name not in private_fields]
+    else:
+        fields = _get_fields(cls)
```
Comment (lijok) [inline strawberry/type.py:110]: Would've written this as
```py
    fields = _get_fields(cls)
    if private_fields:
        fields = list(filter(lambda x: x.origin_name not in private_fields, fields))
```
but mypy was complaining about it
```sh
../../../../../../Users/user/strawberry/strawberry/type:109: error: Unsupported right operand type for in ("Optional[List[str]]") (diff)
```

File: tests/schema/test_private_field.py [hunk 2]
```diff
@@ -23,6 +23,25 @@ class Query:
...
+        name: str
+        age: int
+
+    definition = Query._type_definition
+
+    assert definition.name == "Query"
+    assert len(definition.fields) == 1
+
+    assert definition.fields[0].name == "name"
+    assert definition.fields[0].type == str
```
Comment (BryceBeagle) [inline tests/schema/test_private_field.py:38]: I would add a check here that verifies that `age` is _not_ one of the fields on the `TypeDefinition`

File: strawberry/type.py (hunk 1 above)
Comment (BryceBeagle) [reply strawberry/type.py:110]:
```suggestion
    fields = _get_fields(cls)
    if private_fields:
        fields = [i for i in fields if i.origin_name not in private_fields]
```

Comment (BryceBeagle) [reply strawberry/type.py:110]: Not sure why the `list(filter())` doesn't work, but I think this is a bit cleaner

**Review by BryceBeagle — CHANGES_REQUESTED** [review]
I'm always a bit hesitant to keep tacking on more arguments to the function signatures for our types. I do feel like there's gotta be a cleaner way to get this done. 

I'll let @patrick91 give his opinions here, though

File: strawberry/type.py (hunk 1 above)
Comment (BryceBeagle) [reply strawberry/type.py:110]: Not sure why the list(filter()) doesn't work, but I think this is a bit cleaner

**Comment by eimantas-gecas-axomic** [comment]
> I'm always a bit hesitant to keep tacking on more arguments to the function signatures for our types. I do feel like there's gotta be a cleaner way to get this done.
> 
> I'll let @patrick91 give his opinions here, though

I agree
It's not urgent on my end to get a solution for this in either, we can sleep on this for a while

**Comment by lijok** [comment]
> I'm always a bit hesitant to keep tacking on more arguments to the function signatures for our types. I do feel like there's gotta be a cleaner way to get this done.
> 
> I'll let @patrick91 give his opinions here, though

I agree
It's not urgent to get a solution for this on my end, we can sleep on this

**Comment by patrick91** [comment]
@lijok I think we should work on a plugin for PyCharm instead of changing our code, I don't have bandwidth for that at this time, but maybe in future :)

What do you think? Happy to close this PR for now?

**Comment by lijok** [comment]
> @lijok I think we should work on a plugin for PyCharm instead of changing our code, I don't have bandwidth for that at this time, but maybe in future :)
> 
> What do you think? Happy to close this PR for now?

Yeah, that makes sense
Happy to close this

**Pull request closed without merging**